README.md
OS X RAM Disk
================

Need really fast Java IDE or browser? Then keep reading.

This app will create a [RAM disk](http://en.wikipedia.org/wiki/RAM_drive) in OS-X with specified size to
store the apps cache in RAM, which is known as SSD optimization - reducing disk IO or making browsing the web and
programming using IntelliJ more enjoyable.

Supported apps (you can add yours):

* [IntelliJ Idea 14/15] (https://www.jetbrains.com/idea/download/)
* [Google Chrome] (https://support.google.com/chrome/answer/95346?hl=en)
* [Google Canary] (https://www.google.com/chrome/browser/canary.html)
* Safari
* iTunes
* [Android studio] (http://developer.android.com/sdk/index.html)
* [WebShtorm] (https://www.jetbrains.com/webstorm/)
* [Clion] (https://www.jetbrains.com/clion/)
* [AppCode] (https://www.jetbrains.com/objc/)
* [Xcode] (https://developer.apple.com/xcode/downloads/)
* [your_app_goes_here]

The IntelliJ Idea (or JetBrains IDEs) are really fast after this. Be aware that for the large code base you will
need to have larger RAM Disk. I don't have exact numbers, sorry, it can vary because of many factors. Mine 
machine have 16GB - it works very good for small code bases.

If you observing performance degradation - revise how much memory you are using and may be adding more can help.
By default script will create RAM disk of 1/4 size of your RAM.
If you need to change the size - edit `startupRAMDiskandCacheMover.sh` header section. The RAM disk works with hibernate
option - you don't need to worry about this part.

Have something to discuss? 
[![Join the chat at https://gitter.im/zafarella/OSX-RAMDisk](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/zafarella/OSX-RAMDisk?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/458/badge)](https://bestpractices.coreinfrastructure.org/projects/458)

If you have any issues (compatibility etc) - I am glad to have them reported in the [issues](https://github.com/zafarella/OSX-RAMDisk/issues) tab.

Compatibility
============
Works on

* MAC OS X 10.10.2 Yosemite
* MAC OS X 10.11 EI Capitan

> Note that you have to re-run the script in order to get the ram disk back after machine restart.
> Currently it does not place it on startup - I'm working on it.

Give it a try before installing
===============================
```bash
$ curl -o startupRAMDiskandCacheMover.sh https://raw.githubusercontent.com/zafarella/OSX-RAMDisk/master/Contents/MacOS/startupRAMDiskandCacheMover.sh &&
chmod +x startupRAMDiskandCacheMover.sh &&
./startupRAMDiskandCacheMover.sh
```
or
```
git clone git@github.com:zafarella/OSX-RAMDisk.git &&
OSX-RAMDisk/Contents/MacOS/startupRAMDiskandCacheMover.sh
```
Jetbrains IDE notes (IntelliJ)
===================
In order to speed up the compilation and responsiveness of the IDE you can
change "Project compiler output" and point it to ram disk:
```
echo /Volumes/ramdisk/${USER}/compileroutput
```
and in your project/module settings (Command+down) point to this directory.
You can also set this [setting by default](https://www.jetbrains.com/idea/help/accessing-default-settings.html) 
for your projects at the main menu, choose `File | Other Settings | Default Settings`

In addition to above you can tune jvm with following [flags](https://gist.github.com/zafarella/43bc260c3c0cdc34f109) 
`vim /Applications/IntelliJ\ IDEA\ 15.app/Contents/bin/idea.vmoptions`

Installation
============
Do not use it now - the startup script does not work yet - work in progress
```
git clone git@github.com:zafarella/OSX-RAMDisk.git
cd OSX-RAMDiskix
make install
```

Manual Installation
------------------
```
cp OSXRamDisk.plist ~/Library/LaunchAgents
cp startupRAMDiskandCacheMover.sh /usr/local/bin
# note - it will close Chrome, safari idea
/usr/local/bin/startupRAMDiskandCacheMover.sh
```

Uninstall
============
Run `make uninstall`
or manually do following

Close the chrome, idea or any other application you configured to use ram disk.
```
   rm /usr/local/bin/startupRAMDiskandCacheMover.sh
   launchctl unload -w ~/Library/LaunchAgents/OSXRamDisk.plist
   rm ~/Library/LaunchAgents/OSXRamDisk.plist
```

Alternatives
============
If you are Linux user use
* https://github.com/graysky2/profile-sync-daemon



CHANGELOG.MD
* 1.0
  * initial release with tested on OS X script 10.10.2 Yosemite



Contents/MacOS/createRAMDiskandMoveCaches.sh
#!/usr/bin/env bash

# This is about to create a RAM disk in OS X and move the apps caches into it
# to increase performance of those apps.
# Performance gain is very significat, particularly for browsers and
# especially for IDE
#
# Drawbacks and risks are that if RAM disk becomes full - performance will degrate
# significantly. 
#
# USE AT YOUR OWN RISK.
#



# The RAM amount you want to allocate for RAM disk. One of
# 1024 2048 3072 4096 5120 6144
# todo: set default value to 1/4 of RAM
ramfs_size_mb=4096
mount_point=/Volumes/ramdisk
ramfs_size_sectors=$((${ramfs_size_mb}*1024*1024/512))
ramdisk_device=`hdid -nomount ram://${ramfs_size_sectors}`
USERRAMDISK="$mount_point/$USER"

mk_ram_disk()
{
    # unmount if exists the RAM disk and mounts if doesn't
    umount -f ${mount_point}
    newfs_hfs -v 'ramdisk' ${ramdisk_device}
    mkdir -p ${mount_point}
    mount -o noatime -t hfs ${ramdisk_device} ${mount_point}

    # Hide RAM disk - we don't really need it to be annoiyng in finder.
    # comment out should you need it.
    hide_ramdisk
}

# ------------------------------------------------------
# Application which needs the cache to be moved to RAM
# add yours at the end.
# -------------------------------------------------------

# Google Chrome Cache
move_chrome_cache()
{
    /bin/mkdir -p /tmp/Google/Chrome
    /bin/mv ~/Library/Caches/Google/Chrome/* /tmp/Google/Chrome/
    /bin/mkdir -pv ${USERRAMDISK}/Google/Chrome/Default
    /bin/mv /tmp/Google/Chrome/ ${USERRAMDISK}/Google/Chrome
    /bin/ln -v -s ${USERRAMDISK}/Google/Chrome/Default ~/Library/Caches/Google/Chrome/Default
}

# Chrome Canary Cache
move_chrome_chanary_cache()
{
    /bin/rm -rf ~/Library/Caches/Google/Chrome\ Canary/*
    /bin/mkdir -p ${USERRAMDISK}/Google/Chrome\ Canary/Default
    /bin/ln -s ${USERRAMDISK}/Google/Chrome\ Canary/Default ~/Library/Caches/Google/Chrome\ Canary/Default
}

# Safari Cache
move_safari_cache()
{
    /bin/rm -rf ~/Library/Caches/com.apple.Safari
    /bin/mkdir -p ${USERRAMDISK}/Apple/Safari
    /bin/ln -s ${USERRAMDISK}/Apple/Safari ~/Library/Caches/com.apple.Safari
}

# iTunes Cache
move_itunes_cache()
{
    /bin/rm -rf ~/Library/Caches/com.apple.iTunes
    /bin/mkdir -pv ${USERRAMDISK}/Apple/iTunes
    /bin/ln -v -s ${USERRAMDISK}/Apple/iTunes ~/Library/Caches/com.apple.iTunes
}

# Intellij Idea
move_idea_cache()
{
   # make a backup of config - will need it when uninstalling
   cp -f /Applications/IntelliJ\ IDEA\ 14.app/Contents/bin/idea.properties /Applications/IntelliJ\ IDEA\ 14.app/Contents/bin/idea.properties.back
   # Idea will create those dirs
   echo "idea.system.path=$USERRAMDISK/Idea" >> /Applications/IntelliJ\ IDEA\ 14.app/Contents/bin/idea.properties
   echo "idea.log.path=$USERRAMDISK/Idea/logs" >> /Applications/IntelliJ\ IDEA\ 14.app/Contents/bin/idea.properties
}

# Intellij Idea
move_ideace_cache()
{
   # todo add other versions support and CE edition
   # make a backup of config - will need it when uninstalling
   cp -f /Applications/IntelliJ\ IDEA\ 14\ CE.app/Contents/bin/idea.properties /Applications/IntelliJ\ IDEA\ 14\ CE.app/Contents/bin/idea.properties.back
   # Idea will create those dirs
   echo "idea.system.path=$USERRAMDISK/Idea" >> /Applications/IntelliJ\ IDEA\ 14\ CE.app/Contents/bin/idea.properties
   echo "idea.log.path=$USERRAMDISK/Idea/logs" >> /Applications/IntelliJ\ IDEA\ 14\ CE.app/Contents/bin/idea.properties
}

# todo Android Studio
move_android_studio_cache()
{
    echo "moving Android Studio cache";
    echo "Not implemented"
}

# Closes passed as arg app
close_app()
{
    osascript -e "quit app \"${1}\"" 
}

# Open an application
open_app()
{
     osascript -e "tell app \"${1}\" to activate" 
}

# Hide RamDisk directory
hide_ramdisk()
{
    /usr/bin/chflags hidden ${mount_point}
}

# -----------------------------------------------------------------------------------
# The entry point
# -----------------------------------------------------------------------------------
# Let's close the apps we moving caches for in case they are running.
close_app "IntelliJ Idea 14"
close_app "IntelliJ Idea 14 CE"
close_app "Google Chrome"
close_app "Safari"
close_app "iTunes"
# and create our RAM disk
mk_ram_disk
# move the caches
move_chrome_cache
move_safari_cache
move_idea_cache
move_ideace_cache
move_itunes_cache
echo "All good - I have done my job"
#
# open_app "Google Chrome"
# -----------------------------------------------------------------------------------


Contents/MacOS/startupRAMDiskandCacheMover.sh
#!/usr/bin/env bash -x

#
# Copyright Zafar Khaydarov
#
# This is about to create a RAM disk in OS X and move the apps caches into it
# to increase performance of those apps. Performance gain is very significant,
# particularly for browsers and especially for IDEs like IntelliJ Idea.
#
# Drawbacks and risks are that if RAM disk becomes full - performance will degrade
# significantly - huge amount of paging will happen.
#
# USE AT YOUR OWN RISK. PLEASE NOTE IT WILL NOT CHECK FOR CORRUPTED FILES
# IF YOUR RAM IS BROKEN - DO NOT USE IT.
#

# The RAM amount you want to allocate for RAM disk. One of
# 1024 2048 3072 4096 5120 6144 7168 8192
# By default will use 1/4 of your RAM

ramfs_size_mb=$(sysctl hw.memsize | awk '{print $2;}')
ramfs_size_mb=$((${ramfs_size_mb} / 1024 / 1024 / 4))

mount_point=/Volumes/ramdisk
ramfs_size_sectors=$((${ramfs_size_mb}*1024*1024/512))
ramdisk_device=`hdid -nomount ram://${ramfs_size_sectors}`
USERRAMDISK="$mount_point/${USER}"

MSG_MOVE_CACHE=". Do you want me to move its cache?"
MSG_PROMPT_FOUND="I found "

#
# Checks for the user response.
#
user_response()
{
    echo -ne $@ "[Y/n]  "
    read -r response

    case ${response} in
        [yY][eE][sS]|[yY]|"")
            true
            ;;
        [nN][oO]|[nN])
            false
            ;;
        *)
            user_response $@
            ;;
    esac
}

#
# Closes passed as arg app by name
#
close_app()
{
    osascript -e "quit app \"${1}\""
}

#
# Creates RAM Disk.
#
mk_ram_disk()
{
    # unmount if exists the RAM disk and mounts if doesn't
    umount -f ${mount_point}
    newfs_hfs -v 'ramdisk' ${ramdisk_device}
    mkdir -p ${mount_point}
    mount -o noatime -t hfs ${ramdisk_device} ${mount_point}

    echo "created RAM disk."
    # Hide RAM disk - we don't really need it to be annoiyng in finder.
    # comment out should you need it.
    hide_ramdisk
    echo "RAM disk hidden"
}

# adds rsync to be executed each 5 min for current user
add_rsync_to_cron()
{
    #todo fixme
    crontab -l | { cat; echo "5 * * * * rsync"; } | crontab -
}

# Open an application
open_app()
{
    osascript -e "tell app \"${1}\" to activate"
}

# Hide RamDisk directory
hide_ramdisk()
{
    /usr/bin/chflags hidden ${mount_point}
}

# Checks that we have
# all required utils before proceeding
check_requirements()
{
    hash rsync 2>/dev/null || { echo >&2 "No rsync has been found.  Aborting. If you use brew install using: 'brew install rsync'"; exit 1; }
    hash newfs_hfs 2>/dev/null || { echo >&2 "No newfs_hfs has been found.  Aborting."; exit 1; }
}

#
# Check existence of the string in a file.
#
check_string_in_file()
{
    if  grep "${1}" "${2}" == 0; then
        return 0;
    else
        return 1;
    fi
}

#
# Check for the flag
#
check_for_flag()
{
    if [ -e ${1} ] ; then
        return 0;
    else
        return 1;
    fi
}

#
# Creates flag indicating the apps cache has been moved.
#
make_flag()
{
    echo "" > /Applications/OSX-RAMDisk.app/${1}
}

# ------------------------------------------------------
# Applications, which needs the cache to be moved to RAM
# Add yours at the end.
# -------------------------------------------------------

#
# Google Chrome Cache
#
move_chrome_cache()
{
    if [ -d "/Users/${USER}/Library/Caches/Google/Chrome" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'Chrome'${MSG_MOVE_CACHE} ; then
            close_app "Google Chrome"
            /bin/mkdir -p /tmp/Google/Chrome
            /bin/mv ~/Library/Caches/Google/Chrome/* /tmp/Google/Chrome/
            /bin/mkdir -pv ${USERRAMDISK}/Google/Chrome/Default
            /bin/mv /tmp/Google/Chrome/ ${USERRAMDISK}/Google/Chrome
            /bin/ln -v -s -f ${USERRAMDISK}/Google/Chrome/Default ~/Library/Caches/Google/Chrome/Default
            /bin/rm -rf /tmp/Google/Chrome
            # and let's create a flag for next run that we moved the cache.
            echo "";
        fi
    else
        echo "No Google Chrome folder has been found. Skipping."
    fi
}

#
# Chrome Canary Cache
#
move_chrome_chanary_cache()
{
    if [ -d "/Users/${USER}/Library/Caches/Google/Chrome Canary" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'Chrome Canary'${MSG_MOVE_CACHE} ; then
            close_app "Chrome Canary"
            /bin/rm -rf ~/Library/Caches/Google/Chrome\ Canary/*
            /bin/mkdir -p ${USERRAMDISK}/Google/Chrome\ Canary/Default
            /bin/ln -s ${USERRAMDISK}/Google/Chrome\ Canary/Default ~/Library/Caches/Google/Chrome\ Canary/Default
        fi
    fi
}

#
# Safari Cache
#
move_safari_cache()
{
    if [ -d "/Users/${USER}/Library/Caches/com.apple.Safari" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'Safari'${MSG_MOVE_CACHE}; then
            close_app "Safari"
            /bin/rm -rf ~/Library/Caches/com.apple.Safari
            /bin/mkdir -p ${USERRAMDISK}/Apple/Safari
            /bin/ln -s ${USERRAMDISK}/Apple/Safari ~/Library/Caches/com.apple.Safari
            echo "Moved Safari cache."
        fi
    fi
}

#
# iTunes Cache
#
move_itunes_cache()
{
    if [ -d "/Users/${USER}/Library/Caches/com.apple.iTunes" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'iTunes'${MSG_MOVE_CACHE} ; then
            close_app "iTunes"
            /bin/rm -rf /Users/${USER}/Library/Caches/com.apple.iTunes
            /bin/mkdir -pv ${USERRAMDISK}/Apple/iTunes
            /bin/ln -v -s ${USERRAMDISK}/Apple/iTunes ~/Library/Caches/com.apple.iTunes
            echo "Moved iTunes cache."
        fi
    fi
}

#
# Intellij Idea
#
# fixme - what if the version is not 14?
move_idea_cache()
{
    if [ -d "/Applications/IntelliJ IDEA 14.app" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'IntelliJ IDEA 14'${MSG_MOVE_CACHE} ; then
            close_app "IntelliJ Idea 14"
            # make a backup of config - will need it when uninstalling
            cp -f /Applications/IntelliJ\ IDEA\ 14.app/Contents/bin/idea.properties /Applications/IntelliJ\ IDEA\ 14.app/Contents/bin/idea.properties.back
            # Idea will create those dirs
            echo "idea.system.path=${USERRAMDISK}/Idea" >> /Applications/IntelliJ\ IDEA\ 14.app/Contents/bin/idea.properties
            echo "idea.log.path=${USERRAMDISK}/Idea/logs" >> /Applications/IntelliJ\ IDEA\ 14.app/Contents/bin/idea.properties
            echo "Moved IntelliJ cache."
        fi
    fi


    if [ -d "/Applications/IntelliJ IDEA 15.app" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'IntelliJ IDEA 15'${MSG_MOVE_CACHE} ; then
            close_app "IntelliJ Idea 15"
            # make a backup of config - will need it when uninstalling
            cp -f /Applications/IntelliJ\ IDEA\ 15.app/Contents/bin/idea.properties /Applications/IntelliJ\ IDEA\ 15.app/Contents/bin/idea.properties.back
            # Idea will create those dirs
            echo "idea.system.path=${USERRAMDISK}/Idea" >> /Applications/IntelliJ\ IDEA\ 15.app/Contents/bin/idea.properties
            echo "idea.log.path=${USERRAMDISK}/Idea/logs" >> /Applications/IntelliJ\ IDEA\ 15.app/Contents/bin/idea.properties
            echo "Moved IntelliJ 15 cache."
        fi
    fi
}

#
# Intellij Idea Community Edition
#
move_ideace_cache()
{
    # todo add other versions support and CE edition
    if [ -d "/Applications/IntelliJ IDEA 14 CE.app" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'IntelliJ IDEA CE 14'${MSG_MOVE_CACHE} ; then
            close_app "IntelliJ Idea 14 CE"
            # make a backup of config - will need it when uninstalling
            cp -f /Applications/IntelliJ\ IDEA\ 14\ CE.app/Contents/bin/idea.properties /Applications/IntelliJ\ IDEA\ 14\ CE.app/Contents/bin/idea.properties.back
            # Idea will create those dirs
            echo "idea.system.path=${USERRAMDISK}/Idea" >> /Applications/IntelliJ\ IDEA\ 14\ CE.app/Contents/bin/idea.properties
            echo "idea.log.path=${USERRAMDISK}/Idea/logs" >> /Applications/IntelliJ\ IDEA\ 14\ CE.app/Contents/bin/idea.properties
            echo "Moved IntelliJ cache."
        fi
    fi

    if [ -d "/Applications/IntelliJ IDEA 15 CE.app" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'IntelliJ IDEA CE 15'${MSG_MOVE_CACHE} ; then
            close_app "IntelliJ Idea 14 CE"
            # make a backup of config - will need it when uninstalling
            cp -f /Applications/IntelliJ\ IDEA\ 15\ CE.app/Contents/bin/idea.properties /Applications/IntelliJ\ IDEA\ 15\ CE.app/Contents/bin/idea.properties.back
            # Idea will create those dirs
            echo "idea.system.path=${USERRAMDISK}/Idea" >> /Applications/IntelliJ\ IDEA\ 15\ CE.app/Contents/bin/idea.properties
            echo "idea.log.path=${USERRAMDISK}/Idea/logs" >> /Applications/IntelliJ\ IDEA\ 15\ CE.app/Contents/bin/idea.properties
            echo "Moved IntelliJ cache."
        fi
    fi
}

#
# Creates intelliJ intermediate output folder
# to be used by java/scala projects.
#
create_intermediate_folder_for_intellij_projects()
{
    [ -d /Volumes/ramdisk/${USER}/compileroutput ] || mkdir -p /Volumes/ramdisk/${USER}/compileroutput
}

#
# Android Studio
#
move_android_studio_cache()
{
    if [ -d "/Applications/Android Studio.app" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'Android Studio'${MSG_MOVE_CACHE} ; then
            echo "moving Android Studio cache";
            close_app "Android Studio"
            # make a backup of config - will need it when uninstalling
            cp -f /Applications/Android\ Studio.app/Contents/bin/idea.properties /Applications/Android\ Studio.app/Contents/bin/idea.properties.back
            # Idea will create those dirs
            echo "idea.system.path=${USERRAMDISK}/AndroidStudio" >> /Applications/Android\ Studio.app/Contents/bin/idea.properties
            echo "idea.log.path=${USERRAMDISK}/AndroidStudio/logs" >> /Applications/Android\ Studio.app/Contents/bin/idea.properties
            echo "Moved Android cache."
        fi
    fi
}

#
# Clion
#
move_clion_cache()
{
    if [ -d "/Applications/Clion.app" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'Clion'${MSG_MOVE_CACHE} ; then
            echo "moving Clion cache";
            close_app "Clion"
            # make a backup of config - will need it when uninstalling
            cp -f /Applications/Clion.app/Contents/bin/idea.properties /Applications/Clion.app/Contents/bin/idea.properties.back
            # Idea will create those dirs
            echo "idea.system.path=${USERRAMDISK}/Clion" >> /Applications/Clion.app/Contents/bin/idea.properties
            echo "idea.log.path=${USERRAMDISK}/Clion/logs" >> /Applications/Clion.app/Contents/bin/idea.properties
            echo "Moved Clion cache."
        fi
    fi
}

#
# AppCode - ios
#
move_appcode_cache()
{
    if [ -d "/Applications/AppCode.app" ]; then
        if user_response ${MSG_PROMPT_FOUND} 'AppCode'${MSG_MOVE_CACHE} ; then
            echo "moving AppCode cache";
            close_app "AppCode"
            # make a backup of config - will need it when uninstalling
            cp -f /Applications/AppCode.app/Contents/bin/idea.properties /Applications/AppCode.app/Contents/bin/idea.properties.back
            # Need to create those dirs
            echo "idea.system.path=${USERRAMDISK}/AppCode" >> /Applications/AppCode.app/Contents/bin/idea.properties
            echo "idea.log.path=${USERRAMDISK}/AppCode/logs" >> /Applications/AppCode.app/Contents/bin/idea.properties
            mkdir -p ${USERRAMDISK}/AppCode/logs
            echo "Moved AppCode cache."
        fi
    fi
}

#
# Xcode - ios
#
move_xcode_cache()
{
    if [ -d "/Applications/Xcode.app" ]; then
        if user_response ${MSG_PROMT_FOUND} 'Xcode'${MSG_MOVE_CACHE} ; then
            echo "moving XCode cache..";
            echo "deleting ~/Library/Developer/Xcode/DerivedData"

            /bin/rm -rvf ~/Library/Developer/Xcode/DerivedData
            /bin/mkdir -pv ${USERRAMDISK}/Apple/Xcode
            /bin/ln -v -s ${USERRAMDISK}/Apple/Xcode /Users/${USER}/Library/Developer/Xcode/DerivedData
            echo "Moved Xcode cache."
        fi
    fi
}


# -----------------------------------------------------------------------------------
# The entry point
# -----------------------------------------------------------------------------------
main() {
    check_requirements
    # and create our RAM disk
    mk_ram_disk
    # move the caches
    move_chrome_cache
    move_safari_cache
    move_idea_cache
    move_ideace_cache
    # create intermediate folder for intellij projects output
    create_intermediate_folder_for_intellij_projects
    move_itunes_cache
    move_android_studio_cache
    move_clion_cache
    move_appcode_cache
    move_xcode_cache
    echo "echo use \"/Volumes/ramdisk/${USER}/compileroutput\" for intelliJ project output directory."
    echo "All good - I have done my job. Your apps should fly."
}

main "$@"
# -----------------------------------------------------------------------------------
